File: doc/codes/rust/examples/api/compute/pattern/group.rs
use anyhow::Result;

use autd3_rs::geometry::{Autd3, Geometry, TransducerGroups, offset};
use autd3_rs::units::{m, mm, s};
use autd3_rs::value::Phase;
use autd3_rs_pattern::{focus, group, group_compute, set_phase, wavelength};
use autd3_rs_pattern_holo::{AmplitudeTarget, GspatOption, NalgebraBackend, Pa, gspat};

// ANCHOR: api
// ANCHOR: compute
#[derive(Clone, Copy, PartialEq, Eq)]
enum Side {
    Left,
    Right,
}
// ANCHOR_END: compute
// ANCHOR_END: api

fn main() -> Result<()> {
    let geometry = Geometry::new(vec![Autd3::default()]);

    let left = geometry.phase_buffer();
    let mut right = geometry.phase_buffer();
    set_phase(Phase::PI, &mut right);
    let mut dst = geometry.phase_buffer();
    let center = geometry.center();
    // ANCHOR: api
    let groups = TransducerGroups::new(&geometry, |device, tr| {
        Some(if device.position(tr).x < center.x {
            Side::Left
        } else {
            Side::Right
        })
    });
    group(
        &geometry,
        &groups,
        |side| match side {
            Side::Left => &left,
            Side::Right => &right,
        },
        Phase::ZERO,
        &mut dst,
    );
    // ANCHOR_END: api

    let wavelength = wavelength(340.0 * m / s);
    let foci = [AmplitudeTarget {
        point: center + offset(-30.0 * mm, 0.0 * mm, 150.0 * mm),
        amplitude: 5e3 * Pa,
    }];
    let target = center + offset(40.0 * mm, 0.0 * mm, 150.0 * mm);
    let mut phases = geometry.phase_buffer();
    let mut intensities = geometry.intensity_buffer();
    // ANCHOR: compute
    group_compute(
        &geometry,
        &groups,
        |side, mask, phases, intensities| match side {
            Side::Left => gspat(
                &NalgebraBackend,
                &geometry,
                &foci,
                wavelength,
                &GspatOption {
                    mask,
                    ..Default::default()
                },
                phases,
                intensities,
            ),
            Side::Right => {
                focus(&geometry, target, wavelength, phases);
                Ok(())
            }
        },
        &mut phases,
        &mut intensities,
    )?;
    // ANCHOR_END: compute
    Ok(())
}

File: doc/codes/python/examples/api/compute/pattern/group.py
from enum import Enum, auto

import numpy as np

from autd3.geometry import Autd3, Geometry
from autd3.units import m, s
from autd3.value import Phase
from autd3_pattern import (
    IntensityBuffer,
    PhaseBuffer,
    TransducerGroups,
    TransducerMask,
    focus,
    group,
    group_compute,
    set_phase,
    wavelength,
)
from autd3_pattern_holo import AmplitudeTarget, GspatOption, Pa, gspat


# ANCHOR: api
# ANCHOR: compute
class Side(Enum):
    LEFT = auto()
    RIGHT = auto()
# ANCHOR_END: compute
# ANCHOR_END: api


geometry = Geometry([Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0])])

left = geometry.phase_buffer()
right = geometry.phase_buffer()
set_phase(Phase(0x80), right)
dst = geometry.phase_buffer()
center = geometry.center()

# ANCHOR: api
groups = TransducerGroups(
    geometry,
    lambda device, tr: Side.LEFT if device.position(tr)[0] < center[0] else Side.RIGHT,
)
group(geometry, groups, {Side.LEFT: left, Side.RIGHT: right}, dst)
# ANCHOR_END: api

wl = wavelength(340 * m / s)
foci = [AmplitudeTarget(point=center + np.array([-30.0, 0.0, 150.0]), amplitude=5e3 * Pa)]
target = center + np.array([40.0, 0.0, 150.0])
phases = geometry.phase_buffer()
intensities = geometry.intensity_buffer()


# ANCHOR: compute
def compute(side: Side, mask: TransducerMask, phases: PhaseBuffer, intensities: IntensityBuffer) -> None:
    if side is Side.LEFT:
        gspat(geometry, foci, wl, GspatOption(mask=mask), phases, intensities)
    else:
        focus(geometry, target, wl, phases)


group_compute(geometry, groups, compute, phases, intensities)
# ANCHOR_END: compute
